snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+            "orderby": self._visit_orderby(query.orderby),
+            "limitby": self._visit_limitby(query.limitby),
+            "limit": self._visit_limit(query.limit),
+            "offset": self._visit_offset(query.offset),
+            "granularity": self._visit_granularity(query.granularity),
+            "totals": self._visit_totals(query.totals),
+        }
+
+        return self._combine(query, returns)
+
+    @abstractmethod
+    def _combine(
+        self,
+        query: Query,
+        returns: MutableMapping[str, QVisited],
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_dataset(self, dataset: str) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_match(self, match: Entity) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_select(
+        self, select: Optional[List[Union[Column, Function]]]
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_groupby(
+        self, groupby: Optional[List[Union[Column, Function]]]
+    ) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_where(self, where: Optional[List[Condition]]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_having(self, having: Optional[List[Condition]]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_orderby(self, orderby: Optional[List[OrderBy]]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_limitby(self, limitby: Optional[LimitBy]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_limit(self, limit: Optional[Limit]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_offset(self, offset: Optional[Offset]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_granularity(self, granularity: Optional[Granularity]) -> QVisited:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _visit_totals(self, granularity: Optional[Totals]) -> QVisited:
+        raise NotImplementedError
+
+
+class Printer(QueryVisitor[str]):
+    def __init__(self, pretty: bool = False) -> None:
+        self.translator = Translation()
+        self.pretty = pretty
+
+    def _combine(self, query: Query, returns: MutableMapping[str, str]) -> str:
+        clauses = [
+            "match",
+            "select",
+            "groupby",
+            "where",
+            "having",
+            "orderby",
+            "limitby",
+            "limit",
+            "offset",
+            "granularity",
+            "totals",
+        ]
+        separator = "\n" if self.pretty else " "
+        formatted = separator.join([returns[c] for c in clauses if returns[c]])
+        if self.pretty:
+            formatted = f"-- DATASET: {returns['dataset']}\n{formatted}"
```

Comment:
Do we actually parse comments in SnQL ?  It would be nice to keep the query executable when we pretty print it.

Reply:
We don't, although it would be trivial to add.